File: tools/cabana/dbc/dbcfile.cc
#include "tools/cabana/dbc/dbcfile.h"

#include <QFile>
#include <QFileInfo>
#include <QRegularExpression>

DBCFile::DBCFile(const QString &dbc_file_name) {
  QFile file(dbc_file_name);
  if (file.open(QIODevice::ReadOnly)) {
    name_ = QFileInfo(dbc_file_name).baseName();
    filename = dbc_file_name;
    parse(file.readAll());
  } else {
    throw std::runtime_error("Failed to open file.");
  }
}

DBCFile::DBCFile(const QString &name, const QString &content) : name_(name), filename("") {
  parse(content);
}

bool DBCFile::save() {
  assert(!filename.isEmpty());
  return writeContents(filename);
}

bool DBCFile::saveAs(const QString &new_filename) {
  filename = new_filename;
  return save();
}

bool DBCFile::writeContents(const QString &fn) {
  QFile file(fn);
  if (file.open(QIODevice::WriteOnly)) {
    return file.write(generateDBC().toUtf8()) >= 0;
  }
  return false;
}

void DBCFile::updateMsg(const MessageId &id, const QString &name, uint32_t size, const QString &node, const QString &comment) {
  auto &m = msgs[id.address];
  m.address = id.address;
  m.name = name;
  m.size = size;
  m.transmitter = node.isEmpty() ? DEFAULT_NODE_NAME : node;
  m.comment = comment;
}

cabana::Msg *DBCFile::msg(uint32_t address) {
  auto it = msgs.find(address);
  return it != msgs.end() ? &it->second : nullptr;
}

cabana::Msg *DBCFile::msg(const QString &name) {
  auto it = std::find_if(msgs.begin(), msgs.end(), [&name](auto &m) { return m.second.name == name; });
  return it != msgs.end() ? &(it->second) : nullptr;
}

cabana::Signal *DBCFile::signal(uint32_t address, const QString &name) {
  auto m = msg(address);
  return m ? (cabana::Signal *)m->sig(name) : nullptr;
}

void DBCFile::parse(const QString &content) {
  msgs.clear();

  int line_num = 0;
  QString line;
  cabana::Msg *current_msg = nullptr;
  int multiplexor_cnt = 0;
  bool seen_first = false;
  QTextStream stream((QString *)&content);

  while (!stream.atEnd()) {
    ++line_num;
    QString raw_line = stream.readLine();
    line = raw_line.trimmed();

    bool seen = true;
    try {
      if (line.startsWith("BO_ ")) {
        multiplexor_cnt = 0;
        current_msg = parseBO(line);
      } else if (line.startsWith("SG_ ")) {
        parseSG(line, current_msg, multiplexor_cnt);
      } else if (line.startsWith("VAL_ ")) {
        parseVAL(line);
      } else if (line.startsWith("CM_ BO_")) {
        parseCM_BO(line, content, raw_line, stream);
      } else if (line.startsWith("CM_ SG_ ")) {
        parseCM_SG(line, content, raw_line, stream);
      } else {
        seen = false;
      }
    } catch (std::exception &e) {
      throw std::runtime_error(QString("[%1:%2]%3: %4").arg(filename).arg(line_num).arg(e.what()).arg(line).toStdString());
    }

    if (seen) {
      seen_first = true;
    } else if (!seen_first) {
      header += raw_line + "\n";
    }
  }

  for (auto &[_, m] : msgs) {
    m.update();
  }
}

cabana::Msg *DBCFile::parseBO(const QString &line) {
  static QRegularExpression bo_regexp(R"(^BO_ (?<address>\w+) (?<name>\w+) *: (?<size>\w+) (?<transmitter>\w+))");

  QRegularExpressionMatch match = bo_regexp.match(line);
  if (!match.hasMatch())
    throw std::runtime_error("Invalid BO_ line format");

  uint32_t address = match.captured("address").toUInt();
  if (msgs.count(address) > 0)
    throw std::runtime_error(QString("Duplicate message address: %1").arg(address).toStdString());

  // Create a new message object
  cabana::Msg *msg = &msgs[address];
  msg->address = address;
  msg->name = match.captured("name");
  msg->size = match.captured("size").toULong();
  msg->transmitter = match.captured("transmitter").trimmed();
  return msg;
}

void DBCFile::parseCM_BO(const QString &line, const QString &content, const QString &raw_line, const QTextStream &stream) {
  static QRegularExpression msg_comment_regexp(R"(^CM_ BO_ *(?<address>\w+) *\"(?<comment>(?:[^"\\]|\\.)*)\"\s*;)");

  QString parse_line = line;
  if (!parse_line.endsWith("\";")) {
    int pos = stream.pos() - raw_line.length() - 1;
    parse_line = content.mid(pos, content.indexOf("\";", pos));
  }
  auto match = msg_comment_regexp.match(parse_line);
  if (!match.hasMatch())
    throw std::runtime_error("Invalid message comment format");

  if (auto m = (cabana::Msg *)msg(match.captured("address").toUInt()))
    m->comment = match.captured("comment").trimmed().replace("\\\"", "\"");
}

void DBCFile::parseSG(const QString &line, cabana::Msg *current_msg, int &multiplexor_cnt) {
  static QRegularExpression sg_regexp(R"(^SG_ (\w+) *: (\d+)\|(\d+)@(\d+)([\+|\-]) \(([0-9.+\-eE]+),([0-9.+\-eE]+)\) \[([0-9.+\-eE]+)\|([0-9.+\-eE]+)\] \"(.*)\" (.*))");
  static QRegularExpression sgm_regexp(R"(^SG_ (\w+) (\w+) *: (\d+)\|(\d+)@(\d+)([\+|\-]) \(([0-9.+\-eE]+),([0-9.+\-eE]+)\) \[([0-9.+\-eE]+)\|([0-9.+\-eE]+)\] \"(.*)\" (.*))");

  if (!current_msg)
    throw std::runtime_error("No Message");

  int offset = 0;
  auto match = sg_regexp.match(line);
  if (!match.hasMatch()) {
    match = sgm_regexp.match(line);
    offset = 1;
  }
  if (!match.hasMatch())
    throw std::runtime_error("Invalid SG_ line format");

  QString name = match.captured(1);
  if (current_msg->sig(name) != nullptr)
    throw std::runtime_error("Duplicate signal name");

  cabana::Signal s{};
  if (offset == 1) {
    auto indicator = match.captured(2);
    if (indicator == "M") {
      ++multiplexor_cnt;
      // Only one signal within a single message can be the multiplexer switch.
      if (multiplexor_cnt >= 2)
        throw std::runtime_error("Multiple multiplexor");

      s.type = cabana::Signal::Type::Multiplexor;
    } else {
      s.type = cabana::Signal::Type::Multiplexed;
      s.multiplex_value = indicator.mid(1).toInt();
    }
  }
  s.name = name;
  s.start_bit = match.captured(offset + 2).toInt();
  s.size = match.captured(offset + 3).toInt();
  s.is_little_endian = match.captured(offset + 4).toInt() == 1;
  s.is_signed = match.captured(offset + 5) == "-";
  s.factor = match.captured(offset + 6).toDouble();
  s.offset = match.captured(offset + 7).toDouble();
  s.min = match.captured(8 + offset).toDouble();
  s.max = match.captured(9 + offset).toDouble();
  s.unit = match.captured(10 + offset);
  s.receiver_name = match.captured(11 + offset).trimmed();
  current_msg->sigs.push_back(new cabana::Signal(s));
}

void DBCFile::parseCM_SG(const QString &line, const QString &content, const QString &raw_line, const QTextStream &stream) {
  static QRegularExpression sg_comment_regexp(R"(^CM_ SG_ *(\w+) *(\w+) *\"((?:[^"\\]|\\.)*)\"\s*;)");

  QString parse_line = line;
  if (!parse_line.endsWith("\";")) {
    int pos = stream.pos() - raw_line.length() - 1;
    parse_line = content.mid(pos, content.indexOf("\";", pos));
  }
  auto match = sg_comment_regexp.match(parse_line);
  if (!match.hasMatch())
    throw std::runtime_error("Invalid CM_ SG_ line format");

  if (auto s = signal(match.captured(1).toUInt(), match.captured(2))) {
    s->comment = match.captured(3).trimmed().replace("\\\"", "\"");
  }
}

void DBCFile::parseVAL(const QString &line) {
  static QRegularExpression val_regexp(R"(VAL_ (\w+) (\w+) (\s*[-+]?[0-9]+\s+\".+?\"[^;]*))");

  auto match = val_regexp.match(line);
  if (!match.hasMatch())
    throw std::runtime_error("invalid VAL_ line format");

  if (auto s = signal(match.captured(1).toUInt(), match.captured(2))) {
    QStringList desc_list = match.captured(3).trimmed().split('"');
    for (int i = 0; i < desc_list.size(); i += 2) {
      auto val = desc_list[i].trimmed();
      if (!val.isEmpty() && (i + 1) < desc_list.size()) {
        auto desc = desc_list[i + 1].trimmed();
        s->val_desc.push_back({val.toDouble(), desc});
      }
    }
  }
}

QString DBCFile::generateDBC() {
  QString dbc_string, comment, val_desc;
  for (const auto &[address, m] : msgs) {
    const QString transmitter = m.transmitter.isEmpty() ? DEFAULT_NODE_NAME : m.transmitter;
    dbc_string += QString("BO_ %1 %2: %3 %4\n").arg(address).arg(m.name).arg(m.size).arg(transmitter);
    if (!m.comment.isEmpty()) {
      comment += QString("CM_ BO_ %1 \"%2\";\n").arg(address).arg(QString(m.comment).replace("\"", "\\\""));
    }
    for (auto sig : m.getSignals()) {
      QString multiplexer_indicator;
      if (sig->type == cabana::Signal::Type::Multiplexor) {
        multiplexer_indicator = "M ";
      } else if (sig->type == cabana::Signal::Type::Multiplexed) {
        multiplexer_indicator = QString("m%1 ").arg(sig->multiplex_value);
      }
      dbc_string += QString(" SG_ %1 %2: %3|%4@%5%6 (%7,%8) [%9|%10] \"%11\" %12\n")
                        .arg(sig->name)
                        .arg(multiplexer_indicator)
                        .arg(sig->start_bit)
                        .arg(sig->size)
                        .arg(sig->is_little_endian ? '1' : '0')
                        .arg(sig->is_signed ? '-' : '+')
                        .arg(doubleToString(sig->factor))
                        .arg(doubleToString(sig->offset))
                        .arg(doubleToString(sig->min))
                        .arg(doubleToString(sig->max))
                        .arg(sig->unit)
                        .arg(sig->receiver_name.isEmpty() ? DEFAULT_NODE_NAME : sig->receiver_name);
      if (!sig->comment.isEmpty()) {
        comment += QString("CM_ SG_ %1 %2 \"%3\";\n").arg(address).arg(sig->name).arg(QString(sig->comment).replace("\"", "\\\""));
      }
      if (!sig->val_desc.empty()) {
        QStringList text;
        for (auto &[val, desc] : sig->val_desc) {
          text << QString("%1 \"%2\"").arg(val).arg(desc);
        }
        val_desc += QString("VAL_ %1 %2 %3;\n").arg(address).arg(sig->name).arg(text.join(" "));
      }
    }
    dbc_string += "\n";
  }
  return header + dbc_string + comment + val_desc;
}

File: tools/cabana/pycabana/dbc/dbcfile.py
"""DBC file parser and generator for pycabana."""

import re
from pathlib import Path
from typing import Optional
from enum import Enum

from PySide6.QtCore import QObject

from openpilot.tools.cabana.pycabana.dbc.dbc import MessageId

DEFAULT_NODE_NAME = "XXX"


def double_to_string(value: float) -> str:
  """Convert double to string with maximum precision."""
  # Python's default g format with 15 digits (equivalent to double precision)
  result = f"{value:.15g}"
  return result


def num_decimals(num: float) -> int:
  """Count the number of decimal places in a number."""
  string = str(num)
  dot_pos = string.find('.')
  return 0 if dot_pos == -1 else len(string) - dot_pos - 1


def flip_bit_pos(start_bit: int) -> int:
  """Flip bit position for big-endian signals."""
  return 8 * (start_bit // 8) + 7 - start_bit % 8


def update_msb_lsb(sig: 'Signal') -> None:
  """Update MSB and LSB based on endianness."""
  if sig.is_little_endian:
    sig.lsb = sig.start_bit
    sig.msb = sig.start_bit + sig.size - 1
  else:
    sig.lsb = flip_bit_pos(flip_bit_pos(sig.start_bit) + sig.size - 1)
    sig.msb = sig.start_bit


class SignalType(Enum):
  """Signal type enumeration."""
  Normal = 0
  Multiplexed = 1
  Multiplexor = 2


class Signal:
  """Signal definition in a CAN message."""

  def __init__(self) -> None:
    self.type: SignalType = SignalType.Normal
    self.name: str = ""
    self.start_bit: int = 0
    self.msb: int = 0
    self.lsb: int = 0
    self.size: int = 0
    self.factor: float = 1.0
    self.offset: float = 0.0
    self.is_signed: bool = False
    self.is_little_endian: bool = True
    self.min: float = 0.0
    self.max: float = 0.0
    self.unit: str = ""
    self.comment: str = ""
    self.receiver_name: str = ""
    self.val_desc: list[tuple[float, str]] = []
    self.precision: int = 0
    self.multiplex_value: int = 0
    self.multiplexor: Optional[Signal] = None

  def update(self) -> None:
    """Update signal properties after changes."""
    update_msb_lsb(self)
    if not self.receiver_name:
      self.receiver_name = DEFAULT_NODE_NAME

    # Calculate precision
    self.precision = max(num_decimals(self.factor), num_decimals(self.offset))

  def __eq__(self, other: object) -> bool:
    """Check if two signals are equal."""
    if not isinstance(other, Signal):
      return False
    return (self.name == other.name and self.size == other.size and
            self.start_bit == other.start_bit and
            self.msb == other.msb and self.lsb == other.lsb and
            self.is_signed == other.is_signed and self.is_little_endian == other.is_little_endian and
            self.factor == other.factor and self.offset == other.offset and
            self.min == other.min and self.max == other.max and
            self.comment == other.comment and self.unit == other.unit and
            self.val_desc == other.val_desc and
            self.multiplex_value == other.multiplex_value and self.type == other.type and
            self.receiver_name == other.receiver_name)


class Msg:
  """Message definition in a DBC file."""

  def __init__(self) -> None:
    self.address: int = 0
    self.name: str = ""
    self.size: int = 0
    self.comment: str = ""
    self.transmitter: str = ""
    self.sigs: list[Signal] = []
    self.mask: list[int] = []
    self.multiplexor: Optional[Signal] = None

  def add_signal(self, sig: Signal) -> Signal:
    """Add a signal to the message."""
    self.sigs.append(sig)
    self.update()
    return sig

  def update_signal(self, sig_name: str, new_sig: Signal) -> Optional[Signal]:
    """Update an existing signal."""
    s = self.sig(sig_name)
    if s:
      idx = self.sigs.index(s)
      self.sigs[idx] = new_sig
      self.update()
      return new_sig
    return None

  def remove_signal(self, sig_name: str) -> None:
    """Remove a signal from the message."""
    self.sigs = [s for s in self.sigs if s.name != sig_name]
    self.update()

  def sig(self, sig_name: str) -> Optional[Signal]:
    """Get signal by name."""
    for s in self.sigs:
      if s.name == sig_name:
        return s
    return None

  def index_of(self, sig: Signal) -> int:
    """Get index of signal."""
    try:
      return self.sigs.index(sig)
    except ValueError:
      return -1

  def new_signal_name(self) -> str:
    """Generate a new unique signal name."""
    i = 1
    while True:
      new_name = f"NEW_SIGNAL_{i}"
      if self.sig(new_name) is None:
        return new_name
      i += 1

  def update(self) -> None:
    """Update message properties after changes."""
    if not self.transmitter:
      self.transmitter = DEFAULT_NODE_NAME

    self.mask = [0x00] * self.size
    self.multiplexor = None

    # Sort signals
    def sort_key(s: Signal) -> tuple:
      return (s.type != SignalType.Multiplexor, s.multiplex_value, s.start_bit, s.name)

    self.sigs.sort(key=sort_key)

    # Update each signal
    for sig in self.sigs:
      if sig.type == SignalType.Multiplexor:
        self.multiplexor = sig
      sig.update()

      # Update mask
      i = sig.msb // 8
      bits = sig.size
      while 0 <= i < self.size and bits > 0:
        lsb = sig.lsb if (sig.lsb // 8) == i else i * 8
        msb = sig.msb if (sig.msb // 8) == i else (i + 1) * 8 - 1

        sz = msb - lsb + 1
        shift = lsb - (i * 8)

        self.mask[i] |= ((1 << sz) - 1) << shift

        bits -= sz
        i = i - 1 if sig.is_little_endian else i + 1

    # Set multiplexor references
    for sig in self.sigs:
      sig.multiplexor = self.multiplexor if sig.type == SignalType.Multiplexed else None
      if not sig.multiplexor:
        if sig.type == SignalType.Multiplexed:
          sig.type = SignalType.Normal
        sig.multiplex_value = 0

  def get_signals(self) -> list[Signal]:
    """Get all signals in the message."""
    return self.sigs


class DBCFile(QObject):
  """DBC file parser and generator."""

  def __init__(self, name_or_path: str, content: Optional[str] = None) -> None:
    """
    Initialize DBCFile.

    Args:
      name_or_path: Either a file path (if content is None) or a name (if content is provided)
      content: Optional DBC content as string
    """
    super().__init__()

    self.filename: str = ""
    self.name_: str = ""
    self.header: str = ""
    self.msgs: dict[int, Msg] = {}

    if content is not None:
      # Create from name and content
      self.name_ = name_or_path
      self.filename = ""
      self._parse(content)
    else:
      # Load from file
      path = Path(name_or_path)
      if not path.exists():
        raise RuntimeError("Failed to open file.")
      self.name_ = path.stem
      self.filename = str(path)
      with open(path, encoding='utf-8') as f:
        self._parse(f.read())

  def save(self) -> bool:
    """Save to current filename."""
    assert self.filename, "Filename is empty"
    return self._write_contents(self.filename)

  def save_as(self, new_filename: str) -> bool:
    """Save to a new filename."""
    self.filename = new_filename
    return self.save()

  def _write_contents(self, fn: str) -> bool:
    """Write DBC contents to file."""
    try:
      with open(fn, 'w', encoding='utf-8') as f:
        f.write(self.generate_dbc())
      return True
    except Exception:
      return False

  def update_msg(self, msg_id: MessageId, name: str, size: int, node: str, comment: str) -> None:
    """Update or create a message."""
    if msg_id.address not in self.msgs:
      self.msgs[msg_id.address] = Msg()

    m = self.msgs[msg_id.address]
    m.address = msg_id.address
    m.name = name
    m.size = size
    m.transmitter = DEFAULT_NODE_NAME if not node else node
    m.comment = comment

  def remove_msg(self, msg_id: MessageId) -> None:
    """Remove a message."""
    if msg_id.address in self.msgs:
      del self.msgs[msg_id.address]

  def get_messages(self) -> dict[int, Msg]:
    """Get all messages."""
    return self.msgs

  def msg(self, address_or_name: int | str) -> Optional[Msg]:
    """Get message by address or name."""
    if isinstance(address_or_name, int):
      return self.msgs.get(address_or_name)
    else:
      # Search by name
      for m in self.msgs.values():
        if m.name == address_or_name:
          return m
      return None

  def signal(self, address: int, name: str) -> Optional[Signal]:
    """Get signal by address and name."""
    m = self.msg(address)
    return m.sig(name) if m else None

  def name(self) -> str:
    """Get the name of the DBC file."""
    return "untitled" if not self.name_ else self.name_

  def is_empty(self) -> bool:
    """Check if the DBC file is empty."""
    return len(self.msgs) == 0 and not self.name_

  def _parse(self, content: str) -> None:
    """Parse DBC content."""
    self.msgs.clear()

    lines = content.split('\n')
    line_num = 0
    current_msg: Optional[Msg] = None
    multiplexor_cnt = 0
    seen_first = False

    i = 0
    while i < len(lines):
      line_num += 1
      raw_line = lines[i]
      line = raw_line.strip()

      seen = True
      try:
        if line.startswith("BO_ "):
          multiplexor_cnt = 0
          current_msg = self._parse_bo(line)
        elif line.startswith("SG_ "):
          self._parse_sg(line, current_msg, multiplexor_cnt)
        elif line.startswith("VAL_ "):
          self._parse_val(line)
        elif line.startswith("CM_ BO_"):
          i = self._parse_cm_bo(line, lines, i)
        elif line.startswith("CM_ SG_ "):
          i = self._parse_cm_sg(line, lines, i)
        else:
          seen = False
      except Exception as e:
        raise RuntimeError(f"[{self.filename}:{line_num}]{e}: {line}") from e

      if seen:
        seen_first = True
      elif not seen_first:
        self.header += raw_line + "\n"

      i += 1

    # Update all messages
    for m in self.msgs.values():
      m.update()

  def _parse_bo(self, line: str) -> Msg:
    """Parse BO_ line."""
    bo_pattern = r'^BO_ (?P<address>\w+) (?P<name>\w+) *: (?P<size>\w+) (?P<transmitter>\w+)'
    match = re.match(bo_pattern, line)

    if not match:
      raise RuntimeError("Invalid BO_ line format")

    address = int(match.group('address'))
    if address in self.msgs:
      raise RuntimeError(f"Duplicate message address: {address}")

    msg = Msg()
    msg.address = address
    msg.name = match.group('name')
    msg.size = int(match.group('size'))
    msg.transmitter = match.group('transmitter').strip()

    self.msgs[address] = msg
    return msg

  def _parse_cm_bo(self, line: str, lines: list[str], current_idx: int) -> int:
    """Parse CM_ BO_ (message comment) line."""
    parse_line = line

    # Handle multi-line comments
    if not parse_line.endswith('";'):
      # Find the end of the comment
      for j in range(current_idx + 1, len(lines)):
        parse_line += " " + lines[j].strip()
        if '";' in lines[j]:
          current_idx = j
          break

    msg_comment_pattern = r'^CM_ BO_ *(?P<address>\w+) *"(?P<comment>(?:[^"\\]|\\.)*)"\s*;'
    match = re.match(msg_comment_pattern, parse_line)

    if not match:
      raise RuntimeError("Invalid message comment format")

    address = int(match.group('address'))
    if address in self.msgs:
      comment = match.group('comment').strip().replace('\\"', '"')
      self.msgs[address].comment = comment

    return current_idx

  def _parse_sg(self, line: str, current_msg: Optional[Msg], multiplexor_cnt: int) -> None:
    """Parse SG_ (signal) line."""
    if not current_msg:
      raise RuntimeError("No Message")

    # Try normal signal pattern first
    sg_pattern = r'^SG_ (\w+) *: (\d+)\|(\d+)@(\d+)([\+|\-]) \(([0-9.+\-eE]+),([0-9.+\-eE]+)\) \[([0-9.+\-eE]+)\|([0-9.+\-eE]+)\] "(.*)" (.*)'
    match = re.match(sg_pattern, line)
    offset = 0

    # Try multiplexed signal pattern
    if not match:
      sgm_pattern = r'^SG_ (\w+) (\w+) *: (\d+)\|(\d+)@(\d+)([\+|\-]) \(([0-9.+\-eE]+),([0-9.+\-eE]+)\) \[([0-9.+\-eE]+)\|([0-9.+\-eE]+)\] "(.*)" (.*)'
      match = re.match(sgm_pattern, line)
      offset = 1

    if not match:
      raise RuntimeError("Invalid SG_ line format")

    name = match.group(1)
    if current_msg.sig(name) is not None:
      raise RuntimeError("Duplicate signal name")

    sig = Signal()

    # Handle multiplexing
    if offset == 1:
      indicator = match.group(2)
      if indicator == "M":
        multiplexor_cnt += 1
        if multiplexor_cnt >= 2:
          raise RuntimeError("Multiple multiplexor")
        sig.type = SignalType.Multiplexor
      else:
        sig.type = SignalType.Multiplexed
        sig.multiplex_value = int(indicator[1:])

    sig.name = name
    sig.start_bit = int(match.group(offset + 2))
    sig.size = int(match.group(offset + 3))
    sig.is_little_endian = int(match.group(offset + 4)) == 1
    sig.is_signed = match.group(offset + 5) == "-"
    sig.factor = float(match.group(offset + 6))
    sig.offset = float(match.group(offset + 7))
    sig.min = float(match.group(8 + offset))
    sig.max = float(match.group(9 + offset))
    sig.unit = match.group(10 + offset)
    sig.receiver_name = match.group(11 + offset).strip()

    current_msg.sigs.append(sig)

  def _parse_cm_sg(self, line: str, lines: list[str], current_idx: int) -> int:
    """Parse CM_ SG_ (signal comment) line."""
    parse_line = line

    # Handle multi-line comments
    if not parse_line.endswith('";'):
      # Find the end of the comment
      for j in range(current_idx + 1, len(lines)):
        parse_line += " " + lines[j].strip()
        if '";' in lines[j]:
          current_idx = j
          break

    sg_comment_pattern = r'^CM_ SG_ *(\w+) *(\w+) *"((?:[^"\\]|\\.)*)"\s*;'
    match = re.match(sg_comment_pattern, parse_line)

    if not match:
      raise RuntimeError("Invalid CM_ SG_ line format")

    address = int(match.group(1))
    sig_name = match.group(2)
    s = self.signal(address, sig_name)

    if s:
      comment = match.group(3).strip().replace('\\"', '"')
      s.comment = comment

    return current_idx

  def _parse_val(self, line: str) -> None:
    """Parse VAL_ (value description) line."""
    val_pattern = r'VAL_ (\w+) (\w+) (\s*[-+]?[0-9]+\s+".+?"[^;]*)'
    match = re.match(val_pattern, line)

    if not match:
      raise RuntimeError("invalid VAL_ line format")

    address = int(match.group(1))
    sig_name = match.group(2)
    s = self.signal(address, sig_name)

    if s:
      desc_list = match.group(3).strip().split('"')
      for i in range(0, len(desc_list), 2):
        val_str = desc_list[i].strip()
        if val_str and (i + 1) < len(desc_list):
          desc = desc_list[i + 1].strip()
          s.val_desc.append((float(val_str), desc))

  def generate_dbc(self) -> str:
    """Generate DBC file content."""
    dbc_string = ""
    comment = ""
    val_desc = ""

    for address, m in self.msgs.items():
      transmitter = DEFAULT_NODE_NAME if not m.transmitter else m.transmitter
      dbc_string += f"BO_ {address} {m.name}: {m.size} {transmitter}\n"

      if m.comment:
        escaped_comment = m.comment.replace('"', '\\"')
        comment += f'CM_ BO_ {address} "{escaped_comment}";\n'

      for sig in m.get_signals():
        multiplexer_indicator = ""
        if sig.type == SignalType.Multiplexor:
          multiplexer_indicator = "M "
        elif sig.type == SignalType.Multiplexed:
          multiplexer_indicator = f"m{sig.multiplex_value} "

        endian_char = '1' if sig.is_little_endian else '0'
        sign_char = '-' if sig.is_signed else '+'
        receiver = DEFAULT_NODE_NAME if not sig.receiver_name else sig.receiver_name

        dbc_string += (
          f" SG_ {sig.name} {multiplexer_indicator}: {sig.start_bit}|{sig.size}@{endian_char}{sign_char} "
          + f"({double_to_string(sig.factor)},{double_to_string(sig.offset)}) "
          + f"[{double_to_string(sig.min)}|{double_to_string(sig.max)}] "
          + f'"{sig.unit}" {receiver}\n'
        )

        if sig.comment:
          escaped_comment = sig.comment.replace('"', '\\"')
          comment += f'CM_ SG_ {address} {sig.name} "{escaped_comment}";\n'

        if sig.val_desc:
          text_parts = []
          for val, desc in sig.val_desc:
            text_parts.append(f'{int(val)} "{desc}"')
          val_desc += f"VAL_ {address} {sig.name} {' '.join(text_parts)};\n"

      dbc_string += "\n"

    return self.header + dbc_string + comment + val_desc
